From 77adfafe03b4ce07960755052a16ceec1d779a6e Mon Sep 17 00:00:00 2001
From: Thomas Arcila <134677+tarcila@users.noreply.github.com>
Date: Thu, 24 Apr 2025 13:42:03 -0400
Subject: [PATCH] Build workaround

---
 frame/Frame.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/frame/Frame.cpp b/frame/Frame.cpp
index 3c7c3a5..1775b92 100644
--- a/frame/Frame.cpp
+++ b/frame/Frame.cpp
@@ -477,7 +477,7 @@ void Frame::wait() const
 #else
   if (m_future.valid()) {
     m_future.get();
-    this->refDec(helium::RefType::INTERNAL);
+    const_cast<Frame*>(this)->refDec(helium::RefType::INTERNAL);
     if (deviceState()->currentFrame == this)
       deviceState()->currentFrame = nullptr;
   }
-- 
2.49.0

